src/mvba/hash.rs

```diff
@@ -1,8 +1,65 @@
-#[derive(Debug, Clone, Hash, Eq, PartialEq)]
-pub struct Hash32([u8; 32]);
+use thiserror::Error;
 
-impl From<[u8; 32]> for Hash32 {
-    fn from(val: [u8; 32]) -> Hash32 {
+const HASH32_SIZE: usize = 32;
+
+#[derive(Debug, Clone, Eq, PartialEq, Ord, PartialOrd, Hash)]
+pub struct Hash32([u8; HASH32_SIZE]);
+
+#[derive(Error, Debug, Eq, PartialEq)]
+#[error("invalid length: expected: {}, got: {}", .expected, .found)]
+pub struct InvalidLength {
+    expected: usize,
+    found: usize,
+}
+
+impl Hash32 {
+    pub fn calculate(data: &[u8]) -> Self {
+        use tiny_keccak::{Hasher, Sha3};
+
+        let mut sha3 = Sha3::v256();
+        let mut output = [0; 32];
+        sha3.update(data);
+        sha3.finalize(&mut output);
+        Hash32::from_bytes(&output).unwrap()
```

Comment:
Why not call `Hash32::from_fixed_bytes(output)`? and then drop the `from_bytes` constructor if it's not used elsewhere?

As a rule, we don't allow unwraps / panics in non-test code

Reply:
> drop the from_bytes constructor

Done. I dropped `from_fixed_bytes` and keep `from_bytes`. Is this what you mean?

> As a rule, we don't allow unwraps / panics in non-test code

Agree. Here there is no panic. in line number 20, we created the fixed size array and  panic is impossible. Otherwise all caller should handle the result in their code.

Reply:
>  Please note that there is no panic.

I understand that, but why use the unwrap when you don't need to? `output` is already a 32 byte array, we can safely convert that to Hash32 without the unwraps:

```suggestion
        Hash32::from(output)
```

Reply:
Good idea. Thanks